tools/sdk-generation-pipeline/packages/sdk-generation-lib/src/lib/ArtifactUploader.ts

```diff
@@ -0,0 +1,189 @@
+            artifactBlobUploaderContext.azureStorageBlobSasUrl,
+            artifactBlobUploaderContext.azureBlobContainerName
+        );
+        this.language = artifactBlobUploaderContext.language;
+        this.pipelineBuildId = artifactBlobUploaderContext.pipelineBuildId;
+    }
+
+    private validateContext(artifactBlobUploaderContext: ArtifactBlobUploaderContext) {
+        if (!artifactBlobUploaderContext.azureStorageBlobSasUrl) {
+            throw new Error(`Invalid azureStorageBlobSasUrl`);
+        }
+        if (!artifactBlobUploaderContext.azureBlobContainerName) {
+            throw new Error(`Invalid azureBlobContainerName`);
+        }
+        if (!(<any>Object).values(SDK).includes(artifactBlobUploaderContext.language)) {
+            throw new Error(`Invalid language`);
+        }
+        if (!artifactBlobUploaderContext.pipelineBuildId) {
+            throw new Error(`Invalid pipelineBuildId`);
+        }
+    }
+
+    public async uploadSourceCode(generateAndBuildOutputJson: GenerateAndBuildOutput) {
+        for (const p of generateAndBuildOutputJson.packages) {
+            const result = p.result;
+            if (result === 'failed') {
+                logger.warn(`Build ${p.packageName} failed, skipped it`);
+                continue;
+            }
+            const packageName = p.packageName;
+            const packageFolder = p.packageFolder;
+
+            if (packageFolder && fs.existsSync(packageFolder)) {
+                for (const filePath of getFileListInPackageFolder(packageFolder)) {
+                    if (fs.existsSync(path.join(packageFolder, filePath))) {
+                        await this.azureBlobClient.publishBlob(
+                            path.join(packageFolder, filePath),
+                            `${this.pipelineBuildId}/${this.language}/${packageName}/${filePath}`
+                        );
+                    }
+                }
+            }
+        }
+    }
+
+    public async uploadArtifacts(generateAndBuildOutputJson: GenerateAndBuildOutput) {
+        for (const p of generateAndBuildOutputJson.packages) {
+            const result = p.result;
+            if (result === 'failed') {
+                logger.warn(`Build ${p.packageName} failed, skipped it`);
+                continue;
+            }
+            const artifacts = p.artifacts;
+            if (!artifacts) {
+                // artifacts is optional
+                continue;
+            }
+
+            for (const artifact of artifacts) {
+                const artifactName = path.basename(artifact);
+                await this.azureBlobClient.publishBlob(
+                    artifact,
+                    `${this.pipelineBuildId}/${this.language}/${artifactName}`
+                );
+            }
+        }
+    }
+}
+
+export type ArtifactPipelineUploaderContext = {
+    artifactDir: string;
+    language: string;
+};
+
+export class ArtifactPipelineUploader {
```

Comment:
Seems that 'ArtifactPipelineUploader' is just try to copy files into the target artifact folder, it actually is not publish to the pipeline. In the yml, you still need to call 'ArtifactPublisher' task to do it.
If so, it is not actually ArtifactPipelineUploader. Either you combine all the functionality (copy and publish artifact to pipeline) in this class, or remove this un-completed publisher.

Reply:
I decide to use Azure pipeline 'PublishPipelineArtifact' task to publish artifact. But bash don't support parse json well, so I need to write a cli to finish parse json action and the next copy action.  Give it a name ’getArtifactFiles‘ to distinguish with the old cli.